[MM-17816] Add the ability to transition an issue after running `/jira view`

AGMETEOR opened the pull request:
**Summary**

[MM-17816] Add the ability to transition an issue after running `/jira view`

**Ticket Link**
[https://mattermost.atlassian.net/browse/MM-17816](https://mattermost.atlassian.net/browse/MM-17816)

Fixes #422  

[MM-17816]: https://mattermost.atlassian.net/browse/MM-17816

AGMETEOR commented:
![Screen Shot 2020-01-23 at 19 59 44](https://user-images.githubusercontent.com/28563179/73006483-c813c780-3e1b-11ea-86f5-9e8cc2094817.png)
Still improving this. But this is what I have so far.

hanzei commented:
@mickmister Would you please provide some early feedback?

mickmister commented on server/issue_parser.go:
```diff
@@ -28,7 +28,43 @@ func mdKeySummaryLink(issue *jira.Issue) string {
 }
 
-func parseIssue(issue *jira.Issue) []*model.SlackAttachment {
+func mdTransitionActions(client Client, issue *jira.Issue) ([]*model.PostAction, error) {
```
The `md` prefix is meant for functions returning markdown. This is a similar function in that it is returning information to be rendered in the post. @levb thoughts on this?

AGMETEOR replied:
Hello @mickmister , thanks for the review,
> "Looking at the screenshot, I think it makes sense to have a submit button in the post. Does this work along with the dropdown in the Post Action?"

I have observed that clicking on a transition state triggers a request to transition automatically and I was thinking a submit button isn't really necessary.

> "It is unclear what the status of the ticket is in the post. Maybe show the status in the fields above the dropdown?"

I am not sure if the slack attachment can be dynamically changed and maybe I will need some help there. I can also certainly show the issue's status as of the first execution of `/jira view` but after that, unless instead of showing a simple ephemeral post like "*10001 transitioned to Closed*" we show the a post with an attachment again. 

> "Also, does the call to get transitions provide only valid transitions that make sense with the current status?"

I am not sure if there's a way to filter that out at the moment. Like for example if we have "Under review" status that can only transition to "To do"?

mickmister replied:
> I can also certainly show the issue's status as of the first execution of `/jira view`

Yes, this is what I was thinking

> "Also, does the call to get transitions provide only valid transitions that make sense with the current status?"
> > I am not sure if there's a way to filter that out at the moment.

No worries. I was just curious of Jira did it for us. Trying to implement it on our side may be error-prone.

> But what I can do is to filter the actions such that the current issue status is not among them?

Sounds good to me :+1:

Thanks @AGMETEOR!!

AGMETEOR replied:
@mickmister implemented your feedback. I am also still looking for a way to write tests for these changes as the code coverage seems to have slightly fallen.

levb commented:
((ready for review, I assume? @AGMETEOR )

mickmister commented:
@aaronrothschild This will affect the functionality being introduced in the PR `Reuse output of /jira view in the output of /jira create` https://github.com/mattermost/mattermost-plugin-jira/pull/459. If we choose to display the expanded output after `/jira create`, it will have the transition options on the public post. This may or may not be desirable.

mickmister commented:
Looks like a recent PR is causing merge conflicts. Could you please resolve the conflicts @AGMETEOR? Thanks!

AGMETEOR commented:
@mickmister this is what it looks like at the moment. Any ideas on how to get the transition select button side by side with the reporter info?
![Screen Shot 2020-01-31 at 10 50 29](https://user-images.githubusercontent.com/28563179/73521756-e11e0900-4417-11ea-9501-2cb6954ce77e.png)

levb commented:
@AGMETEOR not a huge deal, but since we squash the PRs upon merge, we prefer

mickmister commented on server/issue.go:
```diff
@@ -20,6 +20,56 @@ import (
+		return http.StatusUnauthorized, errors.New("not authorized")
+	}
+
+	issueKey := requestData.Context["issueKey"].(string)
```
If the request is malformed, and `issueKey` is not present, I believe this line will panic. Perhaps we can use a check like `issueKey, ok := ...` to do some error checking. We can expect it is present in general, but we should be as panic-free as possible.

mickmister commented on server/issue.go:
```diff
@@ -20,6 +20,56 @@ import (
+
+	jiraUser, err := plugin.userStore.LoadJIRAUser(ji, mattermostUserId)
+	if err != nil {
+		_ = plugin.API.SendEphemeralPost(mattermostUserId, makePost(jiraBotID, channelID, "Your username is not connected to Jira. Please type `jira connect`."))
```
Doesn't make much sense to me that this would fail after the transition succeeds. Maybe put this block above the one above?

mickmister commented on server/issue.go:
```diff
@@ -20,6 +20,56 @@ import (
+
+	post.AddProp("attachments", attachment)
+
+	_ = plugin.API.SendEphemeralPost(mattermostUserId, post)
```
Can you remove the blank lines between lines 64-68? I think it would be cleaner as one block.

AGMETEOR replied:
@mickmister I worked on this too.

mickmister commented on server/issue_parser.go:
```diff
@@ -34,7 +34,45 @@ func reporterSummary(issue *jira.Issue) string {
+		if transition.Name != issue.Fields.Status.Name {
+			options = append(options, &model.PostActionOptions{
+				Text:  transition.Name,
+				Value: transition.Name,
```
Can we use `transition.ID` here for the `Value`, and convert `transitionJiraIssue` to use the `ID` instead of trying to match the `Name`?

We originally had it this way so the user could use a command `/jira transition (key) Resolved`, but this feature is essentially replacing the `/jira transition` command. cc @aaronrothschild

AGMETEOR replied:
Hello @mickmister wouldn't this be better done in another ticket after `transitionJiraIssue` is converted to use transition `ID` instead?

mickmister replied:
Yes, sounds good to me :+1:

mickmister commented:
> @AGMETEOR not a huge deal, but since we squash the PRs upon merge, we usually merge rather then rebase/force-push

Merging makes it easier to review, as we know what code additions we have not yet reviewed. It's not possible to know exactly what has changed since last review, if the branch is rebased/force-pushed.

aaronrothschild commented:
> @aaronrothschild This will affect the functionality being introduced in the PR `Reuse output of /jira view in the output of /jira create` #459. If we choose to display the expanded output after `/jira create`, it will have the transition options on the public post. This may or may not be desirable.

@AGMETEOR - This is awesome, I've been waiting for this for a while.  Making actions easier to execute on the view card itself is great (assign users, change priority, etc.).  The biggest concern I have is around security.  

To @mickmister 's question - if the transition options are shown as part of the message, anyone in the channel can select a transition state they may not have permission to use.  Do we know if that would be the case?  or would the user who selected the dropdown be the "active jira user" in that moment and possibly be prevented from transitioning to a state they aren't allowed to?

mickmister commented:
@aaronrothschild 

> Would the user who selected the dropdown be the "active jira user" in that moment and possibly be prevented from transitioning to a state they aren't allowed to?

This is correct. However the output of `/jira view` is not public.

Related - there has been discussion about what should happen when we run `/jira create`, specifically in [this thread](
https://github.com/mattermost/mattermost-plugin-jira/pull/459#discussion_r373304280), as it may be affected by this PR depending on if we want to make the expanded output of `/jira create` to be public. I've expressed that I think we should keep the shortened post `Created (issue link)`, and make the expanded output be private to the user that ran `/jira create`. No rush, but we would like your input on what the behavior should be there.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

hanzei commented:
@AGMETEOR I just want to touch bases with you: Do you have any questions about the feedback from @mickmister?

AGMETEOR commented:
> @AGMETEOR I just want to touch bases with you: Do you have any questions about the feedback from @mickmister?

@hanzei I addressed the feedback that was given.

mickmister commented on server/issue.go:
```diff
@@ -28,43 +28,64 @@ func makePost(userId, channelId, message string) *model.Post {
+	var toState string
+	var found bool
+	var ok bool
+	var x interface{}
```
We can remove these var declarations, based on my suggestions below.

AGMETEOR replied:
Worked on this.

mickmister commented on server/issue.go:
```diff
@@ -28,43 +28,64 @@ func makePost(userId, channelId, message string) *model.Post {
 	}
 }
 
+func respondf(plugin *Plugin, status int, e error, mmuserID, jiraBotID, channelID, message string) (int, error) {
```
Can we rename this to something that implies we are returning an error? Mainly because there are no paths that return `nil` for the error return value. Maybe something like `respondErr`?

AGMETEOR replied:
Worked on this.

mickmister commented on server/issue.go:
```diff
@@ -20,6 +20,77 @@ import (
+
+	jiraUser, err := plugin.userStore.LoadJIRAUser(ji, mattermostUserId)
+	if err != nil {
+		return respondf(plugin, http.StatusUnauthorized, err, mattermostUserId, jiraBotID, channelID, "Your username is not connected to Jira. Please type `jira connect`.")
```
Please add the missing slash for the slash command

```suggestion
		return respondf(plugin, http.StatusUnauthorized, err, mattermostUserId, jiraBotID, channelID, "Your username is not connected to Jira. Please type `/jira connect`.")
```

AGMETEOR replied:
Worked on this.

mickmister commented on server/issue.go:
```diff
@@ -28,43 +28,64 @@ func makePost(userId, channelId, message string) *model.Post {
-	plugin := ji.GetPlugin()
-	jiraBotID := plugin.getUserID()
-	channelID := requestData.ChannelId
+	if x, found = requestData.Context["issueKey"]; found {
```
This is looking a bit too verbose now that we're doing some error checking here. Since we are doing a safe type assertion, I think we can leave out the `found` check and just do the `ok` check afterwards.

```go
val := requestData.Context["issueKey"]
issueKey, ok := val.(string)
if !ok {
// Reply that the issue key is missing
}
```

AGMETEOR replied:
@mickmister Worked on this.

mickmister commented on server/issue_test.go:
```diff
@@ -101,3 +105,47 @@ func TestTransitionJiraIssue(t *testing.T) {
+	request.Header.Set("Mattermost-User-Id", "connected_user")
+	w := httptest.NewRecorder()
+	p.ServeHTTP(&plugin.Context{}, w, request)
+	assert.Equal(t, 400, w.Result().StatusCode)
```
If this test isn't really doing anything except asserting a `400` status. There's not much we can do since we currently don't have Jira API mocking done in our tests. 1/5 I suggest removing the test

levb commented:
@aaronrothschild @DHaussermann ping

aaronrothschild commented:
> 
> Related - there has been discussion about what should happen when we run `/jira create`, specifically in [this thread](https://github.com/mattermost/mattermost-plugin-jira/pull/459#discussion_r373304280), as it may be affected by this PR depending on if we want to make the expanded output of `/jira create` to be public. I've expressed that I think we should keep the shortened post `Created (issue link)`, and make the expanded output be private to the user that ran `/jira create`. No rush, but we would like your input on what the behavior should be there.

@mickmister I agree with that sentiment and approach.

DHaussermann commented:
Tested and passed
- Tested on Cloud and Server
- Tested on Browser and desktop 
- Ensured drop down options are valid in the drop down
- Transitions work as expected

Thanks @AGMETEOR This is a huge UX improvement as I think it's very common to be already looking at view when using transition.   


Note for @aaronrothschild on this - We may want a future UX Story on this. Ideally we would either keep the same "view" of the post once the status is modified OR remove the visual elements from the previous post once a new post with a more current view of the ticket is made.
One drawback with the current design is that previous instances of the "view" in the ephemeral post stay visible and you can manipulate the drop-down. This seems like a broader issue with how plugins make ephemeral posts. 
I don't see this as a blocking problem as the current slash command we have works the same way.
